### PR TITLE
Structural changes

### PR DESCRIPTION
**Pull request by dberliner**
This isn't complete enough to merge, but shows how things will be organized and allows me to improve my talk by Monday evening. This doesn't introduce any real functional change, other than how various parts know what functions are available.

Doing a deep dive into how ECM works was disheartening. The gadget layer has a hard-coded single-use glue structure to interface with Ethernet in [u_ether.c](https://elixir.bootlin.com/linux/latest/source/drivers/usb/gadget/function/u_ether.c#L26). Reinventing the wheel at the level required is not practical for a use case that doesn't actually need Ethernet. 

Luckily, I found that the way I am linking is not forbidden, at least not totally. For instance [this driver](https://elixir.bootlin.com/linux/latest/source/drivers/infiniband/sw/siw/siw_qp_tx.c#L340) links against INET net device functions. I wrote a script to automatically find driver->net references and found they mostly look like 
```
    include/net/bluetooth/hci_core.h, line 1111 (as a prototype)
    net/bluetooth/hci_core.c, line 3579 (as a function) 
Referenced in 17 files:
    drivers/bluetooth/bfusb.c, line 704 
    drivers/bluetooth/...
```
There are no USB drivers that talk to the `/net` directory but our use case isn't common. The thing I want to 100% eliminate is static linking from the network driver to the gadget driver. So the high level goals here are

1. No static linking in the `net -> gadget` direction
2. Central header file for `gadget -> net` direction
3. Minimal interface
4. Move the Proxy (the feature that works with packets) out of the USB driver.

**The New Structure**
`net/scm/xscm.c` -> Network device only (no changes yet) 
`net/scm/xscm_proxy.c` -> Interface that is statically called by the USB driver. This parts calls the USB driver through callbacks set during the initial call. (new file mostly taken from `f_scm.c`)
`include/linux/xscm.h` -> SCM packet struts, functions for accessing and setting members.(unchanged, packet functions not implemented yet)
`include/net/xscm.h` -> Exported `xscm_proxy.c` functions, USB callback function struct.  Both the proxy and USB driver include this.  (new file codifying what used to be inline extern definitions)  
`f_scm.c` -> USB driver interface (absolutely no visibility into _what_ is being sent)

It's also worth noting that `include/net/scm.h` is a file reserved for that Unix socket feature (socket-level control message). The feature [isn't obsolete](https://blog.cloudflare.com/know-your-scm_rights/) and isn't going anywhere. ECM is a great analogy to what we are doing, but I can't think of a good replacement for "S." Here's a few ideas:
* Device Socket Management Protocol
* Virtual Socket Ownership (Management, Protocol)
* Host Socket Passthrough (Sharing)
* eXtended Host Connection Interface (ok,maybe xHCI is taken)

### CODE REVIEW

**Comment by drbild** [comment]
> Luckily, I found that the way I am linking is not forbidden, at least not totally. For instance [this driver](https://elixir.bootlin.com/linux/latest/source/drivers/infiniband/sw/siw/siw_qp_tx.c#L340) links against INET net device functions.

Linking is fine against exported functions, which that is example is: https://elixir.bootlin.com/linux/latest/source/net/ipv4/tcp.c#L1103

Exported functions can be dynamically linked against if the driver is compiled as a module.  And statically linked if the driver is compiled into the kernel binary.

If I'm understanding the following proposal correctly,

> `net/scm/xscm.c` -> Network device only (no changes yet)
`net/scm/xscm_proxy.c` -> Interface that is statically called by the USB driver.

it should work if the functions in `xscm_proxy.c` called by the USB driver are marked as exported.  In that case, they don't have to be statically linked to the USB driver, but can be dynamically linked (when the USB driver is compiled as a module, whether in-tree or out-of-tree).

> I can't think of a good replacement for "S." Here's a few ideas:
> 
> Device Socket Management Protocol
> Virtual Socket Ownership (Management, Protocol)
> Host Socket Passthrough (Sharing)

I like any of those, particularly Host Socket Sharing (or Host Socket Passthrough).

If you wanted to keep the "Control Module" part from USB standards, something like "Host Socket Control Module (HSCM)" could work.  Host Socket Sharing is much more descriptive though.

> eXtended Host Connection Interface (ok,maybe xHCI is taken)

Haha!

File: net/scm/xscm_proxy.c [hunk 1]
```diff
@@ -198,7 +198,6 @@ static void xaprc00x_proxy_process_transmit(struct work_struct *work)
 		break;
 	}
 }
-EXPORT_SYMBOL_GPL(scm_proxy_recv_ack);
```
Comment (drbild) [inline net/scm/xscm_proxy.c]: Why do only a subset of the proxy functions need to be exported?

**Comment by dberliner** [comment]
Yes, your understanding is correct

File: net/scm/xscm_proxy.c (hunk 1 above)
Comment (dberliner) [reply net/scm/xscm_proxy.c]: That's all the USB driver needs. It used to be organized that the USB driver figures out what kind of packet it had and calls the appropriate function in the proxy. I moved that function in here to lower the number if unique calls and remove any packet logic from the USB driver. I could export more but I'm not sure of the use.

Comment (drbild) [reply net/scm/xscm_proxy.c]: :+1:

**Comment by dberliner** [comment]
@drbild This is what we can get in before Wednesday. It includes all of the changes discussed earlier, a bug fix and a complete rename from SCM to HSS. There were some other things I wanted to do but don't have time. This is already a fairly large change due to the rename.

**Review by drbild — APPROVED** [review]
LGTM.

**Pull request merged**